### PR TITLE
Implement reading from Main Section of Capture File

### PR DESCRIPTION
**Pull request by dimitry-**
Implement CaptureSectionInputStream for reading from the
main section of the capture file.

Test: run CaptureFileTests

### CODE REVIEW

File: src/CaptureFile/include/CaptureFile/CaptrureSectionInputStream.h [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+#ifndef CAPTURE_FILE_MAIN_SECTION_INPUT_STREAM_H_
+#define CAPTURE_FILE_MAIN_SECTION_INPUT_STREAM_H_
+
+#include "OrbitBase/Result.h"
+#include "capture.pb.h"
+
+namespace orbit_capture_file {
+
+class CaptureSectionInputStream {
```
Comment (florian-kuebler) [inline src/CaptureFile/include/CaptureFile/CaptrureSectionInputStream.h:13]: Could you add a super brief comment for this class?

File: src/CaptureFile/CaptureSectionInputStreamImpl.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+#ifndef CAPTURE_SECTION_INPUT_STREAM_IMPL_H_
+#define CAPTURE_SECTION_INPUT_STREAM_IMPL_H_
+
+#include "CaptureFile/CaptrureSectionInputStream.h"
+#include "FileFragmentInputStream.h"
+#include "OrbitBase/File.h"
+
+namespace orbit_capture_file_internal {
+
+class CaptureSectionInputStreamImpl : public orbit_capture_file::CaptureSectionInputStream {
```
Comment (florian-kuebler) [inline src/CaptureFile/CaptureSectionInputStreamImpl.h:14]: Could you add a brief comment to the class?

File: src/CaptureFile/CaptureSectionInputStreamImpl.cpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+#include "OrbitBase/MakeUniqueForOverwrite.h"
+
+namespace orbit_capture_file_internal {
+using orbit_grpc_protos::ClientCaptureEvent;
+
+ErrorMessageOr<ClientCaptureEvent> CaptureSectionInputStreamImpl::ReadEvent() {
+  uint64_t message_size = 0;
+
+  if (!coded_input_stream_.ReadVarint64(&message_size)) {
+    return file_fragment_input_stream_.GetLastError().value_or(
```
Comment (florian-kuebler) [inline src/CaptureFile/CaptureSectionInputStreamImpl.cpp:16]: It feels to me that I a missing a small bit. Why is `file_fragment_input_stream_` only used for reading the last error?
How does it correlate with the `coded_input_stream_`?
Could you add a comment somewhere to describe the relationship (could be in the class level comment)?

Comment (florian-kuebler) [reply src/CaptureFile/CaptureSectionInputStreamImpl.cpp:16]: Edit: Okay after reading the header again, I see that coded_input_stream_ is backed by the file_fragment_input_stream_. Still think that a brief comment could help.

File: src/CaptureFile/CaptureFileTest.cpp [hunk 4]
```diff
@@ -20,15 +20,127 @@ static constexpr const char* kNotAnAnswerString = "Some odd number, not the answ
...
+
+  auto output_stream_or_error = CaptureFileOutputStream::Create(temp_file_name);
+  ASSERT_TRUE(output_stream_or_error.has_value()) << output_stream_or_error.error().message();
+
+  std::unique_ptr<CaptureFileOutputStream> output_stream =
+      std::move(output_stream_or_error.value());
+
+  EXPECT_TRUE(output_stream->IsOpen());
+
+  orbit_grpc_protos::ClientCaptureEvent event1 =
```
Comment (florian-kuebler) [inline src/CaptureFile/CaptureFileTest.cpp:99]: I guess you could extract the creation of the capture file into a helper method, as you already have the same code above.

File: src/CaptureFile/CaptureSectionInputStreamImpl.cpp (hunk 3 above)
Comment (dimitry-) [reply src/CaptureFile/CaptureSectionInputStreamImpl.cpp:16]: done.

File: src/CaptureFile/CaptureSectionInputStreamImpl.h (hunk 2 above)
Comment (dimitry-) [reply src/CaptureFile/CaptureSectionInputStreamImpl.h:14]: done

File: src/CaptureFile/include/CaptureFile/CaptrureSectionInputStream.h (hunk 1 above)
Comment (dimitry-) [reply src/CaptureFile/include/CaptureFile/CaptrureSectionInputStream.h:13]: done.

File: src/CaptureFile/CaptureSectionInputStreamImpl.h [hunk 5]
```diff
@@ -0,0 +1,32 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef CAPTURE_SECTION_INPUT_STREAM_IMPL_H_
+#define CAPTURE_SECTION_INPUT_STREAM_IMPL_H_
+
+#include "CaptureFile/CaptrureSectionInputStream.h"
```
Comment (antonrohr) [inline src/CaptureFile/CaptureSectionInputStreamImpl.h:8]:
```suggestion
#include "CaptureFile/CaptureSectionInputStream.h"
```

File: src/CaptureFile/CaptureFileTest.cpp (hunk 4 above)
Comment (florian-kuebler) [reply src/CaptureFile/CaptureFileTest.cpp:99]: As discussed offline, lets keep it as is.

**Review by florian-kuebler — APPROVED** [review]
LGTM, thanks!

File: src/CaptureFile/include/CaptureFile/CaptrureSectionInputStream.h [hunk 6]
```diff
@@ -0,0 +1,25 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef CAPTURE_FILE_MAIN_SECTION_INPUT_STREAM_H_
```
Comment (antonrohr) [inline src/CaptureFile/include/CaptureFile/CaptrureSectionInputStream.h:5]:
```suggestion
#ifndef CAPTURE_SECTION_INPUT_STREAM_H_
```

Comment (dpallotti) [inline src/CaptureFile/include/CaptureFile/CaptrureSectionInputStream.h:5]: Adjust to `CAPTURE_FILE_CAPTURE_SECTION_INPUT_STREAM_H_`.

File: src/CaptureFile/include/CaptureFile/CaptrureSectionInputStream.h [hunk 7]
```diff
@@ -0,0 +1,25 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
```
Comment (dpallotti) [inline src/CaptureFile/include/CaptureFile/CaptrureSectionInputStream.h:1]: The filename has a typo (captrure -> capture).

File: src/CaptureFile/CaptureFileTest.cpp (hunk 4 above)
Comment (dpallotti) [reply src/CaptureFile/CaptureFileTest.cpp:99]: Why does using `CHECK` hurt readability of failures? Because you can't use `<<`? Why not use `FAIL_IF`? We can add `FAIL_UNLESS` if you don't want to invert the condition. There's a bit of duplicated code here.

File: src/CaptureFile/CaptureFile.cpp [hunk 8]
```diff
@@ -55,23 +86,39 @@ ErrorMessageOr<void> CaptureFileImpl::Initialize() {
...
 
-ErrorMessageOr<void> CaptureFileImpl::ReadHeaderAndSectionList() {
-  OUTCOME_TRY(header, orbit_base::ReadFullyAtOffset<CaptureFileHeader>(fd_, 0));
-  header_ = header;
-
-  if (std::memcmp(header_.signature.data(), kFileSignature, header_.signature.size()) != 0) {
-    return ErrorMessage{"Invalid file signature"};
+ErrorMessageOr<void> CaptureFileImpl::CalculateCaptureSectionSize() {
+  // Calculate capture section size
+  if (header_.section_list_offset == 0) {
```
Comment (dpallotti) [inline src/CaptureFile/CaptureFile.cpp:98]: This means that there are no additional sections, right?

File: src/CaptureFile/CaptureFile.cpp [hunk 9]
```diff
@@ -188,6 +240,10 @@ ErrorMessageOr<void> orbit_capture_file::CaptureFileImpl::ReadFromSection(
 
   return outcome::success();
 }
+std::unique_ptr<CaptureSectionInputStream> CaptureFileImpl::CreateCaptureSectionInputStream() {
```
Comment (dpallotti) [inline src/CaptureFile/CaptureFile.cpp:243]:
```suggestion
}

std::unique_ptr<CaptureSectionInputStream> CaptureFileImpl::CreateCaptureSectionInputStream() {
```

File: src/CaptureFile/CaptureFileTest.cpp [hunk 10]
```diff
@@ -20,15 +20,127 @@ static constexpr const char* kNotAnAnswerString = "Some odd number, not the answ
...
+    EXPECT_EQ(event_or_error.value().interned_string().intern(), kAnswerString);
+  }
+
+  {
+    auto event_or_error = capture_section->ReadEvent();
+    ASSERT_TRUE(event_or_error.has_value()) << event_or_error.error().message();
+    ASSERT_EQ(event_or_error.value().event_case(), ClientCaptureEvent::kInternedString);
+    EXPECT_EQ(event_or_error.value().interned_string().key(), kNotAnAnswerKey);
+    EXPECT_EQ(event_or_error.value().interned_string().intern(), kNotAnAnswerString);
+  }
```
Comment (antonrohr) [inline src/CaptureFile/CaptureFileTest.cpp:81]: Maybe call `ReadEvent()` a third time. I would expect an error then

File: src/CaptureFile/CaptureSectionInputStreamImpl.cpp [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+#include "OrbitBase/MakeUniqueForOverwrite.h"
+
+namespace orbit_capture_file_internal {
+using orbit_grpc_protos::ClientCaptureEvent;
+
+ErrorMessageOr<ClientCaptureEvent> CaptureSectionInputStreamImpl::ReadEvent() {
+  uint64_t message_size = 0;
+
+  // Note that in case there was an error CodedInputStream does not provide error messages/coded.
```
Comment (dpallotti) [inline src/CaptureFile/CaptureSectionInputStreamImpl.cpp:15]:
```suggestion
  // Note that in case there was an error CodedInputStream does not provide error messages/codes.
```

**Review by dpallotti — APPROVED** [review]

**Review by antonrohr — APPROVED** [review]
lgtm!

File: src/CaptureFile/CaptureFileTest.cpp (hunk 10 above)
Comment (dimitry-) [reply src/CaptureFile/CaptureFileTest.cpp:81]: I initially had this code but then I realized that:
1. File section can have a padding (filled with 0's) it used to align sections to 8bytes boundary.
2. Reading from stream filled with 0's yields empty (default) message - so this actually succeeds in the case where we have an additional section.
3. We now have CaptureFinished message to serve as terminator of a stream. This message is expected to be used in real world scenarios.

In conclusion - reading past CaptureFinished message is actually undefined even though the stream is bound to the capture section (sometimes it ends in reading default message other times it results in error, it depends if there is a padding in the main section).

And we should probably not test undefined behavior.

Comment (antonrohr) [reply src/CaptureFile/CaptureFileTest.cpp:81]: I agree. Thank you for the explanation! I would like this to be documented somewhere. A comment at `ReadEvent()` in the Impl.h file seems a good idea

File: src/CaptureFile/CaptureFile.cpp (hunk 8 above)
Comment (dimitry-) [reply src/CaptureFile/CaptureFile.cpp:98]: yes. added a comment.

File: src/CaptureFile/CaptureFile.cpp (hunk 9 above)
Comment (dimitry-) [reply src/CaptureFile/CaptureFile.cpp:243]: done

File: src/CaptureFile/CaptureFileTest.cpp (hunk 10 above)
Comment (dimitry-) [reply src/CaptureFile/CaptureFileTest.cpp:81]: added a comment to ReadEvent method

File: src/CaptureFile/CaptureFileTest.cpp (hunk 4 above)
Comment (dimitry-) [reply src/CaptureFile/CaptureFileTest.cpp:99]: yep, using FAIL_IF kind of solves it but it relies on the fact that our tests are isolated and if we use another environment which does not use isolation FAIL_IF aborts the executable, not just one test and that means other tests are not executed. It also feels awkward using FAIL_IF in place of ASSERT_*.

I think having a bit of duplication in tests is fine if doing otherwise results in worse readability of failure results.

File: src/CaptureFile/CaptureSectionInputStreamImpl.cpp (hunk 11 above)
Comment (dimitry-) [reply src/CaptureFile/CaptureSectionInputStreamImpl.cpp:15]: done.

File: src/CaptureFile/CaptureSectionInputStreamImpl.h (hunk 5 above)
Comment (dimitry-) [reply src/CaptureFile/CaptureSectionInputStreamImpl.h:8]: done.

File: src/CaptureFile/include/CaptureFile/CaptrureSectionInputStream.h (hunk 7 above)
Comment (dimitry-) [reply src/CaptureFile/include/CaptureFile/CaptrureSectionInputStream.h:1]: done.

File: src/CaptureFile/include/CaptureFile/CaptrureSectionInputStream.h (hunk 6 above)
Comment (dimitry-) [reply src/CaptureFile/include/CaptureFile/CaptrureSectionInputStream.h:5]: done - kept CAPTURE_FILE_ prefix.

Comment (dimitry-) [reply src/CaptureFile/include/CaptureFile/CaptrureSectionInputStream.h:5]: done.
